### PR TITLE
sessionctx/variable: set DEFAULT_VALUE for 'tidb_txn_mode'

### LINKED ISSUE #48492: The default value for `tidb_txn_mode` is different with doc desc
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
SELECT * FROM information_schema.variables_info WHERE variable_name = 'tidb_txn_mode';
```

### 2. What did you expect to see? (Required)

```
mysql> SELECT * FROM information_schema.variables_info WHERE variable_name = 'tidb_txn_mode';
+---------------+----------------+---------------+---------------+-----------+-----------+------------------------+---------+
| VARIABLE_NAME | VARIABLE_SCOPE | DEFAULT_VALUE | CURRENT_VALUE | MIN_VALUE | MAX_VALUE | POSSIBLE_VALUES        | IS_NOOP |
+---------------+----------------+---------------+---------------+-----------+-----------+------------------------+---------+
| tidb_txn_mode | SESSION,GLOBAL | pessimistic   | pessimistic   |      NULL |      NULL | pessimistic,optimistic | NO      |
+---------------+----------------+---------------+---------------+-----------+-----------+------------------------+---------+
1 row in set (0.01 sec)
```

### 3. What did you see instead (Required)
```
mysql> SELECT * FROM information_schema.variables_info WHERE variable_name = 'tidb_txn_mode';
+---------------+----------------+---------------+---------------+-----------+-----------+------------------------+---------+
| VARIABLE_NAME | VARIABLE_SCOPE | DEFAULT_VALUE | CURRENT_VALUE | MIN_VALUE | MAX_VALUE | POSSIBLE_VALUES        | IS_NOOP |
+---------------+----------------+---------------+---------------+-----------+-----------+------------------------+---------+
| tidb_txn_mode | SESSION,GLOBAL |               | pessimistic   |      NULL |      NULL | pessimistic,optimistic | NO      |
+---------------+----------------+---------------+---------------+-----------+-----------+------------------------+---------+
1 row in set (0.01 sec)
```

### 4. What is your TiDB version? (Required)

ref doc link https://docs.pingcap.com/tidb/stable/system-variables#tidb_txn_mode

### PR DESCRIPTION
**Pull request by highpon**
### What problem does this PR solve?

Issue Number: close #48492

Problem Summary:The documentation currently states that the default value for tidb_txn_mode is pessimistic, however there is no default value set for `tidb_txn_mode`.  This PR fixes this problem and sets the default value of `tidb_txn_mode` to pessimistic.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/infoschema/infoschema_test.go [hunk 1]
```diff
@@ -850,3 +851,12 @@ func TestInfoSchemaCreateTableLike(t *testing.T) {
 	require.Equal(t, tblInfo.Indices[0].Name.O, "idx")
 	require.Equal(t, tblInfo.Indices[0].ID, int64(1))
 }
+
+func TestCheckDefaultValueForTidbTxnMode(t *testing.T) {
```
Comment (Defined2014) [inline pkg/infoschema/infoschema_test.go:855]: No need this test, already covered by https://github.com/pingcap/tidb/blob/af44de3abbb25dd1c9864c6dc261a6763e7f7626/tests/integrationtest/t/infoschema/tables.test#L178

Comment (Defined2014) [reply pkg/infoschema/infoschema_test.go:855]: And not sure does this change will affects upgrade from v3.x.

**Comment by lcwangchao** [comment]
@cfzjywxk  PTAL

File: pkg/infoschema/infoschema_test.go (hunk 1 above)
Comment (highpon) [reply pkg/infoschema/infoschema_test.go]: @Defined2014 
Thanks for your review!
I updated test!
fix commit: https://github.com/pingcap/tidb/pull/48500/commits/033b8f9480b869a3a73bb19dd9848d2c7c202466

**Comment by you06** [comment]
The important reason we leave the default value of variable `tidb_txn_mode` empty is for compatibility when upgrading.

* When the value of `tidb_txn_mode` is `optimistic` or "" or there is no such variable in versions <= v3.0.7, the upgraded cluster should keep the global variable `optimistic`.
* When the value of `tidb_txn_mode` is `pessimistic`, upgraded cluster can use `pessimistic` as global variable.

If we change the default value of `tidb_txn_mode`, the above requirement should be satisfied.

* `tidb_txn_mode` is introduced in `version32`, see https://github.com/pingcap/tidb/blob/96c6ea1ca9a89375d951d0049d108c05a95723a5/session/bootstrap.go#L333
* The v3.0.8 is `version38`, see https://github.com/pingcap/tidb/blob/v3.0.8/session/bootstrap.go#L363

So I think we need to:

* Set global value of `tidb_txn_mode` to `optimistic` in [`upgradeToVer32`](https://github.com/pingcap/tidb/blob/6e8df186f51b90d57e29875b03afe317298bc949/pkg/session/bootstrap.go#L1697-L1702)
* Set global value of `tidb_txn_mode` to `optimistic` if `tidb_txn_mode == ""` in [`upgradeToVer38`](https://github.com/pingcap/tidb/blob/6e8df186f51b90d57e29875b03afe317298bc949/pkg/session/bootstrap.go#L1747-L1752)

**Comment by highpon** [comment]
PR update!

**Review by you06 — COMMENTED** [review]
I tested the the suggested version `upgradeToVer38`, it works.

For `upgradeToVer32`, I cannot start v2 cluster with tiup, it's not easy to test :(

File: pkg/session/bootstrap.go [hunk 2]
```diff
@@ -1699,6 +1699,10 @@ func upgradeToVer32(s Session, ver int64) {
 		return
 	}
 	doReentrantDDL(s, "ALTER TABLE mysql.tables_priv MODIFY table_priv SET('Select','Insert','Update','Delete','Create','Drop','Grant', 'Index', 'Alter', 'Create View', 'Show View', 'Trigger', 'References')")
+
+	// Set optimistic to tidb_tx_mode.
+	// Related issue: https://github.com/pingcap/tidb/issues/48492
+	variable.SetSysVar(variable.TiDBTxnMode, variable.OptimisticTxnMode)
```
[the remark this answers is not in the document]
Comment (highpon) [reply pkg/session/bootstrap.go]: fix: https://github.com/pingcap/tidb/pull/48500/commits/8bf52bceba1aaced72101ebb75773e8e179661f0

File: pkg/session/bootstrap.go [hunk 3]
```diff
@@ -1749,6 +1753,12 @@ func upgradeToVer38(s Session, ver int64) {
 		return
 	}
 	doReentrantDDL(s, CreateGlobalPrivTable)
+
+	// Set optimistic to tidb_tx_mode.
+	// Related issue: https://github.com/pingcap/tidb/issues/48492
+	if variable.GetSysVar(variable.TiDBTxnMode).Value == "" {
+		variable.SetSysVar(variable.TiDBTxnMode, variable.OptimisticTxnMode)
```
[the remark this answers is not in the document]
Comment (highpon) [reply pkg/session/bootstrap.go]: fix: https://github.com/pingcap/tidb/pull/48500/commits/8bf52bceba1aaced72101ebb75773e8e179661f0

File: pkg/sessionctx/variable/sysvar.go [hunk 4]
```diff
@@ -2976,6 +2976,11 @@ var SetCharsetVariables = []string{
 const (
 	// MaskPwd is the mask of password for LDAP variables.
 	MaskPwd = "******"
+
+	// PessimiticTxnMode is the name for tidb_txn_mode system variable.
```
Comment (Defined2014) [inline pkg/sessionctx/variable/sysvar.go:2980]:
```suggestion
	// PessimisticTxnMode is the name for tidb_txn_mode system variable.
```

**Comment by you06** [comment]
@highpon sorry for the idea of seetting variables linside `upgradeToVer32` and `upgradeToVer38`, this won't work if we upgrade to a version without this patch before upgrading it to the latest version.

e.g.
- upgrade from v3.0.0 to latest version, this editted version of `upgradeToVer38` will be executed, fine.
- upgrade from v3.0.0 to v4.0.0, then upgrade from v4.0.0 to latest version, in this case, old version of `upgradeToVer38` is executed when upgrading to v4.0.0, then our grading logic is missing.

I'm discussing with @Defined2014 to find out a compatibility solution.

**Comment by you06** [comment]
Because in early versions, `tidb_txn_mode` is not a global variable, if we upgrade from that version, the correspond record will be missing in `mysql.global_variables` table, then the new default value will be used, which breaks the upgrade compatibility.

So our strategy can be:

* For versions `tidb_txn_mode` is not a global variable, generate "optimistic" value for it.
dc469ad4-6f9b-44ea-9bb9-d3e747cd73bd

* For versions `tidb_txn_mode` is a global variable, keep the global value.
e8616ace-176a-43d8-88bd-f43e1ab84526

* For versions >= v3.0.8, the global value is set to "pessimistic" in #14154, keep it.
image

So we need to execute `INSERT HIGH_PRIORITY IGNORE INTO mysql.GLOBAL_VARIABLES VALUES("tidb_txn_mode", "optimistic")` in the latest `upgradeToVerxxx` function(you can add a new one).

File: pkg/sessionctx/variable/sysvar.go (hunk 4 above)
Comment (highpon) [reply pkg/sessionctx/variable/sysvar.go]: fixed: https://github.com/pingcap/tidb/pull/48500/commits/2386c9f620c0a577e7bf47176926fe0166d58c6a

**Comment by highpon** [comment]
PR updated!

File: pkg/session/bootstrap.go [hunk 5]
```diff
@@ -2892,6 +2897,19 @@ func upgradeToVer179(s Session, ver int64) {
 	doReentrantDDL(s, "ALTER TABLE mysql.global_variables MODIFY COLUMN `VARIABLE_VALUE` varchar(16383)")
 }
 
+func upgradeToVer180(s Session, ver int64) {
+	if ver >= version179 {
+		return
+	}
+
+	sql := fmt.Sprintf("INSERT HIGH_PRIORITY INTO %s.%s ("+
```
Comment (Defined2014) [inline pkg/session/bootstrap.go:2905]: Maybe `INSERT HIGH_PRIORITY IGNORE INTO mysql.GLOBAL_VARIABLES VALUES("tidb_txn_mode", "optimistic")` is enough.
No need to update when `VARIABLE_VALUE` is empty. /cc @you06

Comment (highpon) [reply pkg/session/bootstrap.go]: fixed: https://github.com/pingcap/tidb/pull/48500/commits/4e263ed2a1bfcf8b78af13cf9805edad9386135f

**Review by Defined2014 — APPROVED** [review]

**Review by you06 — APPROVED** [review]
LGTM, sorry for late review, because I missed the notice from GitHub.
Can you solve the conflict so we can merge this PR?
@benmaoer PTAL

File: pkg/session/bootstrap.go [hunk 6]
```diff
@@ -1041,6 +1041,10 @@ const (
 	// version 181
 	//   add column `bdr_role` to `mysql.tidb_ddl_job` and `mysql.tidb_ddl_history`.
 	version181 = 181
+  
+  //   version 182
+  //   set tidb_txn_mode to Optimistic when tidb_txn_mode is not set.
+  version182 = 182
```
Comment (Defined2014) [inline pkg/session/bootstrap.go:1047]:
```suggestion
  
	// version 182
	//   set tidb_txn_mode to Optimistic when tidb_txn_mode is not set.
	version182 = 182
```

File: pkg/session/bootstrap.go [hunk 7]
```diff
@@ -1041,6 +1041,10 @@ const (
 	// version 181
 	//   add column `bdr_role` to `mysql.tidb_ddl_job` and `mysql.tidb_ddl_history`.
 	version181 = 181
+
```
Comment (Defined2014) [inline pkg/session/bootstrap.go:1044]:
```suggestion

```

File: pkg/session/bootstrap.go [hunk 8]
```diff
@@ -1041,6 +1041,9 @@ const (
 	// version 181
 	//   add column `bdr_role` to `mysql.tidb_ddl_job` and `mysql.tidb_ddl_history`.
 	version181 = 181
+	// version 182
```
Comment (Defined2014) [inline pkg/session/bootstrap.go:1044]:
```suggestion

	// version 182
```

**Comment by highpon** [comment]
@Defined2014 
Sorry for the late reply.
Thank you for fixing the conflict.

**Comment by Defined2014** [comment]
> @Defined2014 Sorry for the late reply. Thank you for fixing the conflict.

You are welcome, I will ask PM to review this PR later.

**Review by benmaoer — APPROVED** [review]
LGTM

**Comment by easonn7** [comment]
/approve

The default value of the tidb_txn_mode variable has been changed from "null" to "pessimistic," enhancing the clarity of this variable's meaning and ensuring more consistent behavior.

**Pull request merged**